File: AGRA_2/hw04/ap/ap/tarjanAP.py
"""
Algoritmo de Tarjan (AP)

"""

from sys import stdin

G = [[] for _ in range(50000)]
visited = [False for _ in range(50000)]
low = [-1 for _ in range(50000)]
p = [-1 for _ in range(50000)]
apNodes = set()
t, n = int(), int()

def ap():
    global low, visited, p
    for i in range(n):
        low[i] = visited[i] = p[i] = -1

    for i in range(n):
        if visited[i] == -1:
            apAux(i)

def apAux(v):
    global low, visited, p, apNodos, t
    num = 0
    t += 1
    visited[v] = low[v] = t

    for w in G[v]:
        if visited[w] == -1:
            num += 1
            p[w] = v
            apAux(w);
            low[v] = min(low[v], low[w])

            #verificar si es un punto de articulacion
            if p[v] != -1 and low[w] >= visitado[v]:
                apNodos.add(v)
                
        elif w != p[v]:
            low[v] = min(low[v], visitado[w])

    if p[v] == -1 and numHijos > 1:
        apNodos.add(v)

def main():
    global G
    
    n, m = map(int, stdin.readline().split())

    for i in range(m):
        u, v = map(int, stdin.readline().split())
        G[u].append(v)
        G[v].append(u)

    ap()

    if len(apNodos) == 0:
        print("No hay puntos de articulacion")
    else:
        print("Puntos de Articulacion:")
        print(*apNodos)

main()

File: AGRA_2/hw04/ap/ap/tarjanAP.cpp
/*
Algoritmo de Tarjan (AP)
Autor: Carlos Ramirez
Fecha: Abril 3 de 2020

*/

#include <vector>
#include <stack>
#include <iostream>
#include <set>

using namespace std;

int n, t;
vector<vector<int> > adj(50000);
int visitado[50000];
int low[50000];
int padre[50000];
set<int> apNodos;

void apAux(int);

void ap(){
  int i;

  for(i = 0; i < n; i++)
    low[i] = visitado[i] = padre[i] =-1;

  for(i = 0; i < n; i++)
    if(visitado[i] == -1)
      apAux(i);
}

void apAux(int v){
  int w, numHijos = 0;
  visitado[v] = low[v] = ++t;

  for(int i = 0; i < adj[v].size(); i++){
    w = adj[v][i];
    if(visitado[w] == -1){
      numHijos++;
      padre[w] = v;
      apAux(w);
      low[v] = min(low[v], low[w]);

      //verificar si es un punto de articulacion
      if(padre[v] != -1 && low[w] >= visitado[v])
	apNodos.insert(v);
    }
    else if(w != padre[v])
      low[v] = min(low[v], visitado[w]);
  }

  if(padre[v] == -1 && numHijos > 1)
    apNodos.insert(v);
}

int main(){
  int m, i, aux1, aux2;

  cin >> n >> m;

  for(i = 0; i < m; i++){
    cin >> aux1 >> aux2;
    adj[aux1].push_back(aux2);
    adj[aux2].push_back(aux1);
  }

  ap();

  if(apNodos.size() == 0)
    cout << "No hay puntos de articulacion" << endl;
  else{
    cout << "Puntos de Articulacion:";
    for(set<int>::iterator it = apNodos.begin(); it != apNodos.end(); ++it)
      cout << " " << *it;
    cout << endl;
  }
  
  return 0;
}
